superdesk/auth/oauth.py

```diff
@@ -24,53 +24,197 @@
 
 """
 
+import logging
+from datetime import datetime
+from typing import Optional, List
 import superdesk
+from flask import url_for, render_template
+from authlib.integrations.flask_client import OAuth
+from authlib.integrations.requests_client import OAuth2Session
+from authlib.oauth2.rfc6749.wrappers import OAuth2Token
+from bson import ObjectId
+from superdesk.resource import Resource
+from superdesk.services import BaseService
 
-from flask import url_for, session
-from flask_oauthlib.client import OAuth
+from superdesk.auth import auth_user, TEMPLATE
 
-from superdesk.auth import auth_user
 
+logger = logging.getLogger(__name__)
 
 bp = superdesk.Blueprint('oauth', __name__)
+oauth = None
+REFRESH_TOKEN_URL = "https://oauth2.googleapis.com/token"
 
 
 def init_app(app):
+    global oauth
     oauth = OAuth(app)
-    app.client_config['google_auth'] = False
-    if app.config.get('GOOGLE_CLIENT_ID') and app.config.get('GOOGLE_CLIENT_SECRET'):
-        app.client_config['google_auth'] = True
-        configure_google(app, oauth)
-        superdesk.blueprint(bp, app)
 
+    endpoint_name = 'oauth2_token'
+    service = OAuth2TokenService(endpoint_name, backend=superdesk.get_backend())
+    OAuth2TokenResource(endpoint_name, app=app, service=service)
 
-def configure_google(app, oauth):
-    google = oauth.remote_app(
+    app.client_config['google_auth'] = False
+    if app.config.get('GOOGLE_CLIENT_ID') and app.config.get('GOOGLE_CLIENT_SECRET'):
+        app.client_config['google_auth'] = bool(app.config.get('GOOGLE_LOGIN'))
+        if app.config.get('GOOGLE_GMAIL'):
+            extra_scopes = ['https://mail.google.com']
+            refresh = True
+        else:
+            extra_scopes = []
+            refresh = False
+
+        configure_google(app, extra_scopes=extra_scopes, refresh=refresh)
+
+
+class OAuth2TokenResource(Resource):
+    schema = {
+        'identifier': {
```

Comment:
wouldn't `_id` work for that? it has a unique index on it so that's a plus

Reply:
yes it would be better indeed, I need to modify client for that, and I have realised that I don't have ingest provider ID when the `LOG-IN WITH GMAIL` button is pressed, and also that name can be missing at this point (it's mandatory, but when the button is clicked, the ingest is not yet saved).
I'm trying to find a workaround.